instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/engine/BeanAttributeExtractor.java

```diff
@@ -42,31 +44,63 @@ public class BeanAttributeExtractor implements MetricAttributeExtractor {
    * @throws IllegalArgumentException if the attribute name is malformed
    */
   public static BeanAttributeExtractor fromName(String rawName) {
-    if (rawName.isEmpty()) {
-      throw new IllegalArgumentException("Empty attribute name");
+    List<String> segments = splitByDot(rawName);
+    String baseName = segments.remove(0);
+    if (segments.isEmpty()) {
+      return new BeanAttributeExtractor(baseName);
     }
+    return new BeanAttributeExtractor(baseName, segments.toArray(new String[segments.size()]));
+  }
 
-    // Check if a CompositeType value is expected
-    int k = rawName.indexOf('.');
-    if (k < 0) {
-      return new BeanAttributeExtractor(rawName);
-    }
+  /*
+   * Split a given name into segments, assuming that a dot is used to separate the segments.
+   * However, a dot preceded by a backslash is not a separator.
+   */
+  private static List<String> splitByDot(String rawName) {
+    List<String> components = new ArrayList<>();
+    try {
+      StringBuilder currentSegment = new StringBuilder();
+      boolean escaped = false;
+      for (int i = 0; i < rawName.length(); ++i) {
+        char ch = rawName.charAt(i);
+        if (escaped) {
+          // Allow only '\' and '.' to be escaped
+          if (ch != '\\' && ch != '.') {
+            throw new IllegalArgumentException(
+                "Invalid escape sequence in attribute name '" + rawName + "'");
+          }
+          currentSegment.append(ch);
+          escaped = false;
+        } else {
+          if (ch == '\\') {
+            escaped = true;
+          } else if (ch == '.') {
+            // this is a segment separator
+            verifyAndAddNameSegment(components, currentSegment);
+            currentSegment = new StringBuilder();
+          } else {
+            currentSegment.append(ch);
+          }
+        }
+      }
 
-    // Set up extraction from CompositeType values
-    String baseName = rawName.substring(0, k).trim();
-    String[] components = rawName.substring(k + 1).split("\\.");
+      // The returned list is never empty ...
+      verifyAndAddNameSegment(components, currentSegment);
 
-    // sanity check
-    if (baseName.isEmpty()) {
+    } catch (IllegalArgumentException unused) {
+      // Drop the original exception. We have more meaningful context here.
       throw new IllegalArgumentException("Invalid attribute name '" + rawName + "'");
     }
-    for (int j = 0; j < components.length; ++j) {
-      components[j] = components[j].trim();
-      if (components[j].isEmpty()) {
-        throw new IllegalArgumentException("Invalid attribute name '" + rawName + "'");
-      }
+
+    return components;
+  }
+
+  private static void verifyAndAddNameSegment(List<String> segments, StringBuilder candidate) {
+    String newSegment = candidate.toString().trim();
+    if (newSegment.isEmpty()) {
+      throw new IllegalArgumentException();
```

Comment:
Nit: can you add an exception message about an empty name?

Reply:
This exception instance will be caught and ignored by the calling method, which will throw another one with the message quoting the original attribute (complete) name. Do you think it is not sufficient?

Reply:
Ah, I missed that. No, that will be fine 👍